grpc/src/main/java/com/linecorp/armeria/internal/grpc/ArmeriaMessageDeframer.java

```diff
@@ -0,0 +1,503 @@
+            while (pendingDeliveries > 0 && readRequiredBytes()) {
+                switch (state) {
+                    case HEADER:
+                        processHeader();
+                        break;
+                    case BODY:
+                        // Read the body and deliver the message.
+                        processBody();
+
+                        // Since we've delivered a message, decrement the number of pending
+                        // deliveries remaining.
+                        pendingDeliveries--;
+                        break;
+                    default:
+                        throw new IllegalStateException("Invalid state: " + state);
+                }
+            }
+
+            /*
+            * We are stalled when there are no more bytes to process. This allows delivering errors as
+            * soon as the buffered input has been consumed, independent of whether the application
+            * has requested another message.  At this point in the function, either all frames have been
+            * delivered, or unprocessed is empty.  If there is a partial message, it will be inside next
+            * frame and not in unprocessed.  If there is extra data but no pending deliveries, it will
+            * be in unprocessed.
+            */
+            boolean stalled = unprocessed.readableBytes() == 0;
+
+            if (endOfStream && stalled) {
+                boolean havePartialMessage = nextFrame != null && nextFrame.readableBytes() > 0;
+                if (!havePartialMessage) {
+                    listener.endOfStream();
+                    deliveryStalled = false;
+                    return;
+                } else {
+                    // We've received the entire stream and have data available but we don't have
+                    // enough to read the next frame ... this is bad.
+                    throw Status.INTERNAL.withDescription(
+                            DEBUG_STRING + ": Encountered end-of-stream mid-frame").asRuntimeException();
+                }
+            }
+            deliveryStalled = stalled;
+        } finally {
+            inDelivery = false;
+        }
+    }
+
+    /**
+     * Attempts to read the required bytes into nextFrame.
+     *
+     * @return {@code true} if all of the required bytes have been read.
+     */
+    private boolean readRequiredBytes() {
+        if (nextFrame == null) {
+            nextFrame = alloc.compositeBuffer();
+        }
+
+        // Read until the buffer contains all the required bytes.
+        int missingBytes;
+        while ((missingBytes = requiredLength - nextFrame.readableBytes()) > 0) {
+            if (unprocessed.readableBytes() == 0) {
+                // No more data is available.
+                return false;
+            }
+            int toRead = Math.min(missingBytes, unprocessed.readableBytes());
+            nextFrame.addComponent(true, unprocessed.readBytes(toRead));
```

Comment:
Actually went back to readBytes since everything broke, guess the discardReadComponents made a copy here necessary. Let me know if there's a better pattern.